components/observations/submitObservation.ts

```diff
@@ -0,0 +1,115 @@
+import axios from 'axios';
+import * as FileSystem from 'expo-file-system';
+import {isUndefined, omit} from 'lodash';
+import md5 from 'md5';
+
+import {ObservationFormData} from 'components/observations/ObservationFormData';
+import {AvalancheCenterID, MediaItem, mediaItemSchema, Observation} from 'types/nationalAvalancheCenter';
+import {apiDateString} from 'utils/date';
+
+const extensionToMimeType = (extension: string) => {
+  switch (extension.toLowerCase()) {
+    case 'jpeg':
+    case 'jpg':
+      return 'image/jpeg';
+    case 'png':
+      return 'image/png';
+    default:
+      throw new Error(`Unknown mime type for extension: ${extension}`);
+  }
+};
+
+const imageUploadCachePrefix = 'IMAGE_UPLOAD_CACHE';
+
+// TODO: should put an expiration time on these cache entries and clear them less aggressively.
+export const clearUploadCache = async () => {
+  const keys = await AsyncStorage.getAllKeys();
+  keys.filter(k => k.startsWith(imageUploadCachePrefix)).map(async k => await AsyncStorage.removeItem(k));
+};
+
+const uploadImage = async ({apiPrefix, uri, name, center_id}: {apiPrefix: string; center_id: AvalancheCenterID; uri: string; name: string | undefined}): Promise<MediaItem> => {
+  // This weird use of `slice` is because the version of Hermes that Expo is currently pinned to doesn't support `at()`
+  const filename = uri.split('/').slice(-1)[0];
+  const extension = filename.split('.').slice(-1)[0] || '';
+
+  const base64Data = await FileSystem.readAsStringAsync(uri, {encoding: 'base64'});
+  const payload = omit(
+    {
+      file: `data:${extensionToMimeType(extension)};base64,${base64Data}`,
+      type: 'image',
+      file_name: filename,
+      center_id,
+      forecast_zone_id: [],
+      taken_by: name,
+      access: 'anonymous', // TODO: plumb through use with photo credit / don't use
+      source: 'public',
+      // TODO would be nice to tag images that came from this app, but haven't figured that out yet
+    },
+    isUndefined,
+  );
+
+  // If we've already uploaded this image once, don't do it again.
+  const payloadHash = md5(payload);
+  const imageCacheKey = `${imageUploadCachePrefix}:${payloadHash}`;
+  const cached = await AsyncStorage.getItem(imageCacheKey);
+  if (cached) {
+    console.log(`Image ${uri} has already been uploaded, using cached media item`);
+    try {
+      return Promise.resolve(mediaItemSchema.parse(JSON.parse(cached)));
+    } catch (error) {
+      console.warn(`Unable to load cached image data for ${uri}, uploading it again`);
+      await AsyncStorage.removeItem(imageCacheKey);
+      // fallthrough
+    }
+  }
+
+  const response = await axios.post<MediaItem>(`${apiPrefix}/v2/public/media`, payload, {
+    headers: {
+      // Public API uses the Origin header to determine who's authorized to call it
+      Origin: 'https://nwac.us',
+    },
+  });
+  await AsyncStorage.setItem(imageCacheKey, JSON.stringify(response.data));
```

Comment:
Why couldn't these also be `useMutation`s?

Reply:
They don't play well with hooks: there are from 0-8 of them, and also we need to sequence these calls: the media have to be uploaded before the observation. I think it's cleaner this way.